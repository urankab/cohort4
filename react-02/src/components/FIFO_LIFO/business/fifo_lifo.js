async function getElephant() {
        const proxyurl = 'https://cors-anywhere.herokuapp.com/'
        const url = 'https://elephant-api.herokuapp.com/elephants/random'
        fetch(proxyurl + url)
            .then((resp) => { return resp.json() })
            .then((data) => {
                data.forEach((elephant) => {
                    const { name, sex, species, note, id } = elephant
                    document.getElementById('name').value = name
                    document.getElementById('gender').value = sex
                    document.getElementById('species').value = species
                    document.getElementById('about').value = note
                })
            })
        .catch(()=> console.log("Can't access " + url + " blocked?"))
}

export default getElephant