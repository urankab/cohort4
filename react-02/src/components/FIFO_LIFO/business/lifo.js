export default class Stacks {
    constructor() {
        this.storage = {}
        this.size = 0
        this.lastKey=0
    }

    push(element) {
        this.lastKey++
        this.size++
        element['key'] = this.lastKey
        this.storage[this.size] = element
    }

    pop() {
        let removed = this.storage[this.size]
        delete this.storage[this.size]
        this.size--
        return removed.name
    }

    //Let's us see the last element
    peek() {
        if (this.size >= 1) {
            return this.storage[this.size].name
        }
        else{
            return null
        }
    }
}

