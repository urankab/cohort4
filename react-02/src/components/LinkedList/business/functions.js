class ListNode {
   constructor(element) {
      this.next = null
      this.element = element
      this.amount = amount
   }
   show() {
      return `${this.element} amount: ${this.amount}`
   }
}

// One of the states that must be maintained will be the position we are in the linked list. To that end implement the following functions:
// first ⇒ position to the first node
// last ⇒ position to the last node
// next ⇒ move to the next node
// previous ⇒ backup one node (how are we going to do this?)
// insert ⇒ inserts a new node after the current node (which node will be the current node after the insertion?)
// delete ⇒ delete the current node (which node will be the current node after the deletion?)


class LinkedList {
   constructor() {
      this.head = null;
      this.size = 0.
   }

   insert(element) {
      let node = new ListNode(element)
      let current
      //If list is empty, add the element to make it head
      if (this.head == null) {
         this.head = node
      }
      else {
         current = this.head
         //Iterate to the end of the list
         while (current.next) {
            current = current.next
         }
         //Add node
         current.next = node
      }
      this.size++
   }

   delete(element) {
      let current = this.head
      let prev = null
      //Iterate over list
      while (current != null) {
         //Compare element with current, if found, then remove 
         if (current.element === element) {
            if (prev == null) {
               this.head = current.next
            } else {
               prev.next = current.next
            }
            this.size--
            return current.element
         }
         prev = current
         current = current.ext
      }
      return -1
   }

   next(element) {

   }

   previous(element) {

   }

   first(element) {
      let node = new Node(element)
   }

   last(element) {

   }

   total() {
      return this.size
   }
}

export { ListNode, LinkedList };

//Linked list: Successful nodes are connected by pointers, last node returns null
// Allocates memory as it grows unless arrays which have a fixed size
// A linked list can easily insert or remove nodes from a list without reorganization
// of the entire dat structure

// Drawbacks: Random access of elements is not allowed, nodes must be accessed 
// starting from the first onemptied, so it is slow
// It uses more memory than arrays because of using pointers

//Single linked list: 2 parts: data and a pointer to the next node, terminates at null