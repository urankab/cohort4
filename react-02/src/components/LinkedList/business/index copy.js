class Node {
   constructor(subject, amount, prev, next) {
      this.subject = subject
      this.amount = amount
      this.next = next || null
      this.prev = prev || null
   }
   show() {
      return `Subject: ${this.subject} --- Amount: ${this.amount}`
   }
}

class LinkedList {
   constructor() {
      this.head = null
      this.tail = null
      this.size = 0
   }

   // add to the end of the list / tail
   append(subject, amount) {
      // if list is empty
      if (!this.tail) {
         this.head = this.tail = new Node(subject, amount)
      }
      // if linkedlist has >= one node
      else {
         let oldTail = this.tail
         this.tail = new Node(subject, amount)
         oldTail.next = this.tail
         this.tail.prev = oldTail
      }
      this.size++
   }

   //add to beginning of list / head
   prepend(subject, amount) {
      // if list is empty
      if (!this.head) {
         this.head = this.tail = new Node(subject, amount)
      }
      // if linkedlist has >= one node
      else {
         let oldHead = this.head
         this.head = new Node(subject, amount)
         oldHead.prev = this.head
         this.head.next = oldHead
      }
      this.size++
   }

   deleteHead() {
      // if list is empty (no head)
      if (!this.head) {
         return null
      }
      // if linkedlist has >= 1 node
      else {
         let removedHead = this.head
         // if list has only 1 node left
         if (this.head === this.tail) {
            this.head = this.tail = null
         }
         //if list has >1 node
         else {
            this.head = this.head.next
            this.head.prev = null
         }
         this.size--
         return removedHead.value
      }
   }

   deleteTail() {
      // if list is empty (no tail)
      if (!this.tail) {
         return null
      }
      // if linkedlist has >= one node
      else {
         let removedTail = this.tail
         // if list has only 1 node left
         if (this.head === this.tail) {
            this.tail = this.head = null
         }
         //if list has >1 node
         else {
            this.tail = this.tail.prev
            this.tail.next = null
         }
         this.size--
         return `Deleted ${removedTail.subject}`
      }
   }

   //Search for specific subject and amount
   search(subject, amount) {
      let currentNode = this.head

      while (currentNode) {
         if (currentNode.subject === subject && currentNode.amount === amount) {
            return currentNode
         }
         currentNode = currentNode.next
      }
      return null
   }

   totalAmounts() {
      let currentNode = this.head;
      let total = 0;
      while (currentNode) {
         total += Number(currentNode.amount)
         currentNode = currentNode.next
      }
      return total
   }

   printList() {
      let curr = this.head;
      let str = '';
      while (curr) {
         str += ` -| ${curr.subject} - ${curr.amount} |- `
         curr = curr.next
      }
      return str
   }

   // searchBySubject(subject) {
   //    let currentNode = this.head
   //    let amounts = ''

   //    while (currentNode) {
   //       if (currentNode.subject === subject) {
   //          amounts += currentNode.amount
   //       }
   //       currentNode = currentNode.next
   //    }
   //    return amounts
   // }
}

export { LinkedList, Node }
