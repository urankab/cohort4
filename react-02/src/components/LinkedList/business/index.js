class Node {
   constructor(subject, amount, prev, next) {
      this.subject = subject
      this.amount = amount
      this.next = next || null
      this.prev = prev || null
   }
   show() {
      return `Subject: ${this.subject} --- Amount: ${this.amount}`
   }
}

class LinkedList {
   constructor() {
      this.head = null
      this.tail = null
      this.currentNode = null
      this.size = 0
   }

   first() {
      this.currentNode = this.head
      return this.currentNode
   }

   last() {
      this.currentNode = this.tail
      return this.currentNode
   }

   prev() {
      this.currentNode = this.currentNode.prev
      return this.currentNode
   }

   next() {
      this.currentNode = this.currentNode.next
      return this.currentNode
   }

   insert(subject, amount) {
      //If the list is empty, make it the head and current node
      if (!this.head) {
         this.head = this.tail = new Node(subject, amount)
         this.currentNode = this.head
      }
      //If the current node is the head, make new node the head
      else if (this.currentNode === this.head) {
         let oldHead = this.head
         this.head = new Node(subject, amount)
         oldHead.prev = this.head
         this.head.next = oldHead
         this.currentNode = this.head
      }
      //If there is stuff, add before the currentNode, then make it the new current node
      else {
         let oldCurrent = this.currentNode
         this.currentNode = new Node(subject, amount)
         let temp = oldCurrent.prev; //Temp is old current node's prev
         temp.next = this.currentNode
         this.currentNode.prev = temp
         this.currentNode.next = oldCurrent
         oldCurrent.prev = this.currentNode
      }
      this.size++
   }

   delete() {
      //If list is empty
      if (!this.head) {
         return null
      }
      else {
         let removedNode = this.currentNode
         //If list only has 1 item
         if (this.head === this.tail) {
            this.head = this.tail = null
            console.log('1')
            this.size--
            return `Deleted ${removedNode.subject}`
         }
         //If the CN is the head, return null and let head.next be new head
         else if (removedNode === this.head) {
            this.head = this.head.next
            this.head.prev = null
            this.currentNode = this.head
            console.log('2')
            this.size--
            return `Deleted ${removedNode.subject}`
         }
         //If the CN is the tail, return null and let tail.prev be new tail
         else if (removedNode === this.tail) {
            this.tail = this.tail.prev
            this.tail.next = null
            this.currentNode = this.tail
            console.log('3')
            this.size--
            return `Deleted ${removedNode.subject}`
         }
         else {
            //If current node is inbetween head and tail, new Current is the prev
            let tempPrev = this.currentNode.prev
            let tempNext = this.currentNode.next
            this.currentNode = tempPrev
            tempPrev.next = tempNext
            this.size--
            console.log('4')
            return `Deleted ${removedNode.subject}`
         }
      }
   }

   //Search for specific subject and amount
   search(subject, amount) {
      let currentNode = this.head

      while (currentNode) {
         if (currentNode.subject === subject && currentNode.amount === amount) {
            return currentNode
         }
         currentNode = currentNode.next
      }
      return null
   }

   totalAmounts() {
      let currentNode = this.head;
      let total = 0;
      while (currentNode) {
         total += Number(currentNode.amount)
         currentNode = currentNode.next
      }
      return total
   }

   printList() {
      let curr = this.head;
      let str = '';
      while (curr) {
         // str += ` -| ${curr.subject} - ${curr.amount} |- `
         str += `${curr.subject} - ${curr.amount}\n`
         curr = curr.next
      }
      console.log(str)
      // return str
   }

   // searchBySubject(subject) {
   //    let currentNode = this.head
   //    let amounts = ''

   //    while (currentNode) {
   //       if (currentNode.subject === subject) {
   //          amounts += currentNode.amount
   //       }
   //       currentNode = currentNode.next
   //    }
   //    return amounts
   // }
}

export { LinkedList, Node }
