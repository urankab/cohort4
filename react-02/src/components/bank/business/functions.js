export class Account {
    static lastKey = 0
    constructor(obj) {
        const defaults = { accountName: '', balance: '', key: '' }
        const data = { ...defaults, ...obj }

        this.accountName = data.accountName;
        this.balance = data.balance;
        this.key = data.key;
    }

    newKey() {
        Account.lastKey++;
        this.key = Account.lastKey;
    }

    deposit(valueIn) {
        this.balance += Number(valueIn)
    }

    withdraw(valueOut) {
        this.balance -= Number(valueOut)
    }

    showBalance() {
        return `${this.accountName}: $${this.balance}`;
    }

    showName() {
        return this.accountName;
    }
}

export class AccountController {
    constructor() {
        this.accounts = {}
        this.lastKey = 0;
        this.selectedName = ''
    }

    getDefaults() {
        return new Account({})
    }

    getAccountByKey(key) {
        return this.accounts[key]
    }

    // getAccountNameByKey(thekey) {
    //     for (let key in this.accounts) {
    //         console.log(thekey)
    //         if (Number(thekey) === this.accounts[key].key) {
    //             // this.selectedName = this.accounts[key].accountName
    //             // console.log(this.accounts[key].accountName)
    //             return this.accounts[key].accountName
    //         } else {
    //             return `fail: ${this.accounts[key].key}`
    //         }
    //     }
    // }

    getAccountNameByKey(thekey) {
        for (let key in this.accounts) {
            if ((Number(thekey) === this.accounts[key].key)) {
                return this.accounts[key].accountName
            }
        }
    }

    checkLength() {
        return Object.keys(this.accounts).length
    }

    checkName(nameToCheck) {
        for (let key in this.accounts) {
            if (nameToCheck === this.accounts[key].accountName) {
                return true;
            }
        }
        return false;
    }

    addAccount(account) {
        this.lastKey++
        account.key = this.lastKey;
        this.accounts[account.key] = account
    }

    removeAccount(nameToDelete) {
        for (let key in this.accounts) {
            if (nameToDelete === this.accounts[key].accountName) {
                delete this.accounts[key]
            }
        }
    }

    renameAccount(accountName, newName) {
        for (let key in this.accounts) {
            if (accountName === this.accounts[key].accountName) {
                this.accounts[key].accountName = newName;
            }
        }
    }

    totalBalance() {
        let total = 0
        for (let key in this.accounts) {
            total += Number(this.accounts[key].balance)
        }
        if (this.checkLength() > 0)
            return `$${total}`
    }

    highestAccount() {
        let highestBal = 0;
        let highestName;
        for (let key in this.accounts) {
            if (this.accounts[key].balance > highestBal) {
                let currentName = this.accounts[key].accountName;
                highestBal = Number(this.accounts[key].balance);
                highestName = currentName;
            }
        }
        if (this.checkLength() > 0)
            return `${highestName} - $${highestBal}`
    }

    lowestAccount() {
        let lowest = Number.POSITIVE_INFINITY;
        let lowestName;
        for (let key in this.accounts) {
            let currentName = this.accounts[key].accountName;
            let currentBal = Number(this.accounts[key].balance);
            if (currentBal < lowest) {
                lowestName = currentName;
                lowest = currentBal;
            }
        }
        if (this.checkLength() > 0)
            return `${lowestName} - $${lowest}`
    }

    showAll() {
        let stuff = ''
        for (let key in this.accounts) {
            stuff += `${this.accounts[key].accountName} - $${this.accounts[key].balance}\n`
        }
        return stuff;
    }
}