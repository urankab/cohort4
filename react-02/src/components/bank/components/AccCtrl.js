import React from 'react'

function AccCtrl(props) {
   let options;
   if (props.accounts) {
      options = Object.keys(props.accounts).map(k => {
         const a = props.accounts[k]
         return (
            <option key={a.key} mykey={a.key} value={a.accountName}
            >
               {a.accountName} - ${a.balance}
            </option>
         )
      })
   }

   function focusElement(name) {
      const el = document.querySelector(`[id=${name}]`);
      el.focus();
   }

   function onDeposit() {
      if (props.accLength >= 1) {
         let e = document.getElementById('dropdown')
         let theKey = e.options[e.selectedIndex].getAttribute("mykey")
         let theName = e.options[e.selectedIndex].value

         let amt = document.getElementById('amount').value
         if (theKey != '') {
            if (amt != '') {
               props.acctCtrl.getAccountByKey(theKey).deposit(amt)
               props.userMsg('')
               props.userEditMsg(`Deposited ${amt} to ${theName}`)
               props.updateSummary()
            } else {
               focusElement('amount');
               props.userMsg('')
               props.userEditMsg('Please enter an amount to deposit')
            }
         } else {
            props.userMsg('')
            props.userEditMsg('Select account to deposit')
         }
      } else {
         props.userEditMsg('No accounts to deposit too')
      }
   }

   function onWithdraw() {
      if (props.accLength >= 1) {
         let e = document.getElementById('dropdown')
         let theKey = e.options[e.selectedIndex].getAttribute("mykey")
         let theName = e.options[e.selectedIndex].value

         let amt = document.getElementById('amount').value
         if (theKey != '') {
            if (amt != '') {
               props.acctCtrl.getAccountByKey(theKey).withdraw(amt)
               props.userMsg('')
               props.userEditMsg(`Withdrawed ${amt} to ${theName}`)
               props.updateSummary()
            } else {
               focusElement('amount');
               props.userMsg('')
               props.userEditMsg('Please enter an amount to withdraw')
            }
         } else {
            props.userMsg('')
            props.userEditMsg('Select account to withdraw from')
         }
      } else {
         props.userEditMsg('No accounts to withdraw from')
      }
   }

   function onDelete() {
      if (props.accLength >= 1) {
         let e = document.getElementById('dropdown')
         let theKey = e.options[e.selectedIndex].getAttribute("mykey")
         let theName = e.options[e.selectedIndex].value
         if (theKey != '') {
            props.delete(theKey)
            props.userEditMsg(`Deleted ${theName}`)
         }
      }
      else {
         focusElement('dropdown');
         props.userEditMsg('No accounts to delete')
      }
   }

   function onRename() {
      if (props.accLength >= 1) {
         let renameValue = document.getElementById('renameField').value
         let e = document.getElementById('dropdown')
         let theKey = e.options[e.selectedIndex].getAttribute("mykey")
         let accName = e.options[e.selectedIndex].value
         if (theKey !== '') {
            if (renameValue !== '') {
               props.rename(theKey, renameValue)
               props.userEditMsg(`Renamed ${accName}`)
            } else {
               focusElement('renameField')
               props.userEditMsg('Please enter a new name')
            }
         }
      }
      else {
         props.userEditMsg('No accounts to rename')
      }
   }

   return (
      <div id='box2'>
         <h2 className='boxHeader'>Edit an Account</h2>
         <div className='innerDiv'>
            <p className='msg'>{props.editMsg}</p>
            <label id='selectLabel' htmlFor='dropdown'>Select Account: </label>
            <select id='dropdown' name='dropdown'>
               {options}
            </select>
            <br></br>
            <label htmlFor='amount'>Deposit/Withdraw: $</label>
            <input
               type='number' className='input' id='amount'></input>
            <br></br>
            <button className='btn' id='depBtn' onClick={onDeposit}>Deposit</button>
            <button className='btn' id='withdrawBtn' onClick={onWithdraw}>Withdraw</button>
            <br></br>
            <label htmlFor='renameField'>New Name: </label>
            <input type='text' className='input' id='renameField'></input>
            <button className='btn' id='renameBtn' onClick={onRename}>
               Change Name</button>
            <button className='btn' id='deleteBtn' onClick={onDelete}>
               Delete Account</button>
         </div>
      </div >
   )
}

export default AccCtrl;