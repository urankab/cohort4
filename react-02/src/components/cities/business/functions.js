const url = 'http://localhost:5000/';

async function postData(url = '', data = {}) {
    // Default options are marked with *
    const response = await fetch(url, {
        method: 'POST',     // *GET, POST, PUT, DELETE, etc.
        mode: 'cors',       // no-cors, *cors, same-origin
        cache: 'no-cache',  // *default, no-cache, reload, force-cache, only-if-cached
        credentials: 'same-origin', // include, *same-origin, omit
        headers: {
            'Content-Type': 'application/json'
            // 'Content-Type': 'application/x-www-form-urlencoded',
        },
        redirect: 'follow',         // manual, *follow, error
        referrer: 'no-referrer',    // no-referrer, *client
        body: JSON.stringify(data)  // body data type must match "Content-Type" header
    });

    const json = await response.json();    // parses JSON response into native JavaScript objects
    json.status = response.status;
    json.statusText = response.statusText;

    return json;
}

class City {
    static lastKey = 0;
    constructor(obj) {
        const defaults = {
            name: "", latitude: '', longitude: '', population: "", key: ""
        }
        const data = { ...defaults, ...obj };
        this.name = data.name;
        this.latitude = data.latitude;
        this.longitude = data.longitude;
        this.population = data.population;
        this.key = data.key;
    }

    newKey() {
        City.lastKey++;
        this.key = City.lastKey;
    }

    async movedIn(num) {
        this.population += Number(num);
        await postData(url + 'update', { population: this.population, key: this.key })
    }

    async movedOut(num) {
        this.population -= Number(num);
        await postData(url + 'update', { population: this.population, key: this.key })
    }

    show() {
        return `${this.name}\nLatitude: ${this.latitude}\nLongitude: ${this.longitude}\nPopulation: ${this.population}\nKey: ${this.key}`;
    }

    howBig() {
        if (this.population === 0) {
            return "No Population"
        }
        if (this.population > 0 && this.population <= 100) {
            return "Hamlet: 1-100";
        }
        if (this.population > 100 && this.population < 1000) {
            return "Village: 101-999"
        }
        if (this.population >= 1000 && this.population < 20000) {
            return "Town: 1,000-20,000";
        }
        if (this.population >= 20000 && this.population < 100000) {
            return "Large Town: 20,000-100,000";
        }
        if (this.population >= 100000) {
            return "City: 100,000+";
        }
    }

    whichSphere() {
        if (this.latitude > 0) {
            return 'Northern Hemisphere'
        }
        if (this.latitude === 0) {
            return 'Equater'
        }
        if (this.latitude < 0) {
            return 'Southern Hemisphere'
        }
    }
}

class Community {
    constructor() {
        this.cities = {};
        this.lastKey = 0;
    }

    length() {
        return Object.keys(this.cities).length
    }

    get(key) {
        return this.cities[key]
    }

    getNewCity() {
        return new City({})
    }

    async loadCities() {
        const data = await postData(url + "all");

        //Create a dictionary of cities and keep track of the last key
        const cities = {};
        data.forEach(x => {
            cities[x.key] = x;
            this.lastKey = (x.key > this.lastKey) ? x.key : this.lastKey;
        });

        this.cities = cities;
    }

    async addOrUpdate(city) {
        let theUrl;

        if (city.key) {
            theUrl = url + 'update'
            // city.population = Number(city.population)
        } else {
            theUrl = url + 'add'
            this.lastKey++;
            city.key = this.lastKey;
            // city.population = Number(city.population)
        }
        await postData(theUrl, city);
        this.cities[city.key] = city;
    }

    async delete(thekey) {
        let theUrl;
        if (thekey) {
            theUrl = url + 'delete'
        }
        await postData(theUrl, { key: Number(thekey) });
    }

    getMostNorthern() {
        let latitudeArray = []
        let theName;
        for (let key in this.cities) {
            latitudeArray.push(this.cities[key].latitude)
        }
        let mostNorthLat = Math.max.apply(null, latitudeArray);
        for (let key in this.cities) {
            if (mostNorthLat === this.cities[key].latitude) {
                theName = this.cities[key].name
            }
        }
        if (this.length() > 0) {
            return `${theName} at ${mostNorthLat}°`
        }
    }

    getMostSouthern() {
        let latitudeArray = []
        let theName;
        for (let key in this.cities) {
            latitudeArray.push(this.cities[key].latitude)
        }
        let mostSouthLat = Math.min.apply(null, latitudeArray);
        for (let key in this.cities) {
            if (mostSouthLat === this.cities[key].latitude) {
                theName = this.cities[key].name
            }
        }
        if (this.length() > 0) {
            return `${theName} at ${mostSouthLat}°`
        }
    }

    getTotalPopulation() {
        let total = 0
        for (let key in this.cities) {
            total += Number(this.cities[key].population)
        }
        if (this.length() > 0) {
            return total
        }
    }
}

export default { City, Community, url, postData };