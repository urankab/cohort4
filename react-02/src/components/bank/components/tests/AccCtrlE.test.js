import React from 'react';
import { shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { fireEvent, render, screen } from '@testing-library/react'
import AccCtrl from '../AccCtrl'
import { AccountController } from '../../business/functions'

var Enzyme = require('enzyme');
Enzyme.configure({ adapter: new Adapter() })

describe('AccCtrl Component test', () => {

   it('Snapshot?', () => {
      let wrapper = shallow(<AccCtrl />)
      expect(wrapper).toMatchSnapshot()
   })

   it('Renders the dropdown list and the header', () => {
      const header = 'Edit an Account'

      let wrapper = shallow(<AccCtrl
      />)

      expect(wrapper.find('#dropdown'))
      expect(wrapper.find('h2')).toHaveLength(1)
      expect(wrapper.contains(header)).toBe(true)
   })

   it('Test adding and renaming an account validation', () => {
      const acctCtrl = new AccountController()
      const accounts = acctCtrl.accounts

      acctCtrl.addAccount({ accountName: 'Kitty', balance: 100 })
      acctCtrl.addAccount({ accountName: 'Doggy', balance: 10 })

      const mockRenameClickCallBack = jest.fn()
      const mockUserMsgCallBack = jest.fn()

      let wrapper = shallow(<AccCtrl
         userEditMsg={mockUserMsgCallBack}
      />)

      render(<AccCtrl
         accounts={accounts}
         rename={mockRenameClickCallBack}
      />)

      const a1 = 'Kitty - $100'
      wrapper.find('#renameBtn').simulate('click')
      expect(mockUserMsgCallBack.mock.calls.length).toBe(1)
      expect(mockRenameClickCallBack.mock.calls.length).toBe(0)
      expect(mockUserMsgCallBack.mock.calls[0][0]).toBe('Please enter a new name')
   })

   it('Test adding and renaming an account', () => {
      const acctCtrl = new AccountController()
      const accounts = acctCtrl.accounts

      acctCtrl.addAccount({ accountName: 'Kitty', balance: 100 })
      acctCtrl.addAccount({ accountName: 'Doggy', balance: 10 })

      const mockRenameClickCallBack = jest.fn()
      const mockUserMsgCallBack = jest.fn()

      let wrapper = shallow(<AccCtrl
         userEditMsg={mockUserMsgCallBack}
         rename={mockRenameClickCallBack}
      />)

      render(<AccCtrl
         accounts={accounts}
         rename={mockRenameClickCallBack}
      />)

      updateValue('renameField', 'Cats')
      wrapper.find('#renameBtn').simulate('click')
      expect(mockUserMsgCallBack.mock.calls.length).toBe(1)
      expect(mockRenameClickCallBack.mock.calls.length).toBe(1)
      expect(mockUserMsgCallBack.mock.calls[0][0]).toBe('Renamed Kitty')
      // wrapper.find('#dropdown').simulate('change', { target: { value: 'Doggy' } })
      // // expect(wrapper.find('#dropdown').props().value).toBe('Doggy')
   })

   it('Test renaming a different selection', () => {
      const acctCtrl = new AccountController()
      const accounts = acctCtrl.accounts

      acctCtrl.addAccount({ accountName: 'Kitty', balance: 100 })
      acctCtrl.addAccount({ accountName: 'Doggy', balance: 10 })

      const mockRenameClickCallBack = jest.fn()
      const mockUserMsgCallBack = jest.fn()

      let wrapper = mount(<AccCtrl
         userEditMsg={mockUserMsgCallBack}
         rename={mockRenameClickCallBack}
         accounts={accounts}
      />)

      render(<AccCtrl
         accounts={accounts}
         rename={mockRenameClickCallBack}
      />)

      updateValue('renameField', 'Wolves')

      document.getElementById('dropdown').selectedIndex = '1'
      let e = document.getElementById('dropdown')
      let value = e.options[e.selectedIndex].value
      expect(value).toBe('Doggy')

      wrapper.find('#renameBtn').simulate('click')

      expect(mockRenameClickCallBack.mock.calls.length).toBe(1)
      expect(mockUserMsgCallBack.mock.calls.length).toBe(1)
      expect(mockUserMsgCallBack.mock.calls[0][0]).toBe('Renamed Doggy')
   })

   it('Test deleting an account validation', () => {
      const acctCtrl = new AccountController()
      const accounts = acctCtrl.accounts

      acctCtrl.addAccount({ accountName: 'Kitty', balance: 100 })
      acctCtrl.addAccount({ accountName: 'Doggy', balance: 10 })

      const mockDeleteClickCallBack = jest.fn()
      const mockUserMsgCallBack = jest.fn()

      let wrapper = shallow(<AccCtrl
         userEditMsg={mockUserMsgCallBack}
         delete={mockDeleteClickCallBack}
      />)

      render(<AccCtrl
         accounts={accounts}
         delete={mockDeleteClickCallBack}
      />)

      wrapper.find('#deleteBtn').simulate('click')
      expect(mockUserMsgCallBack.mock.calls.length).toBe(1)
      expect(mockDeleteClickCallBack.mock.calls.length).toBe(1)
      expect(mockUserMsgCallBack.mock.calls[0][0]).toBe('Deleted Kitty')
   })
})

function getValue(name) {
   return document.querySelector(`[id=${name}]`).value
}

function getTextContent(name) {
   return document.getElementById(`${name}`).textContent;
}

function updateValue(name, value) {
   document.querySelector(`[id=${name}]`).value = value;
}