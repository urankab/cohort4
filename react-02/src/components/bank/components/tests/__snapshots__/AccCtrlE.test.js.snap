// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccCtrl Component test Snapshot? 1`] = `ShallowWrapper {}`;
